src/keyrefresh/keyshare.rs

```diff
@@ -0,0 +1,244 @@
+        let public_share = private_share.public_point()?;
+
+        Ok((
+            private_share,
+            KeyUpdatePublic::new(participant, public_share),
+        ))
+    }
+
+    pub(crate) fn sum(participant: ParticipantIdentifier, shares: &[Self]) -> Self {
+        let sum = shares.iter().fold(CurvePoint::IDENTITY, |sum, o| sum + o.X);
+        Self {
+            participant,
+            X: sum,
+        }
+    }
+
+    pub(crate) fn apply(&self, current_pk: &KeySharePublic) -> KeySharePublic {
+        let sum = *current_pk.as_ref() + self.X;
+        KeySharePublic::new(current_pk.participant(), sum)
+    }
+}
+
+impl AsRef<CurvePoint> for KeyUpdatePublic {
+    /// Get the public curvepoint which is the public key share.
+    fn as_ref(&self) -> &CurvePoint {
+        &self.X
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{
+        auxinfo,
+        utils::{k256_order, testing::init_testing},
+    };
+    use rand::rngs::StdRng;
+
+    /// Generate an encryption key pair.
+    fn setup() -> (StdRng, EncryptionKey, DecryptionKey) {
+        let mut rng = init_testing();
+        let pid = ParticipantIdentifier::random(&mut rng);
+        let auxinfo = auxinfo::Output::simulate(&[pid], &mut rng);
+        let dk = auxinfo.private_auxinfo().decryption_key();
+        let pk = auxinfo.find_public(pid).unwrap().pk();
+        assert!(
+            &(k256_order() * 2) < pk.modulus(),
+            "the Paillier modulus is supposed to be much larger than the k256 order"
+        );
+        (rng, pk.clone(), dk.clone())
+    }
+
+    #[test]
+    fn key_update_encryption_works() {
+        let (mut rng, pk, dk) = setup();
+        let rng = &mut rng;
+
+        // Encryption round-trip.
+        let share = KeyUpdatePrivate::random(rng);
+        let encrypted = KeyUpdateEncrypted::encrypt(&share, &pk, rng).expect("encryption failed");
+        let decrypted = encrypted.decrypt(&dk).expect("decryption failed");
+
+        assert_eq!(decrypted, share);
+    }
+
+    #[test]
+    fn key_update_decrypt_out_of_range() {
+        let (mut rng, pk, dk) = setup();
+        let rng = &mut rng;
+
+        // Encrypt invalid shares.
+        for x in [BigNumber::zero(), -BigNumber::one(), k256_order()].iter() {
+            let share = KeyUpdatePrivate { x: x.clone() };
+            let encrypted =
+                KeyUpdateEncrypted::encrypt(&share, &pk, rng).expect("encryption failed");
+            // Decryption reports an error.
+            let decrypt_result = encrypted.decrypt(&dk);
+            assert!(decrypt_result.is_err());
+        }
+    }
+
+    #[test]
+    fn key_update_private_zero_sum_works() {
+        // Random shares do not sum to zero.
+        let rng = &mut init_testing();
+        let mut shares = (0..5)
+            .map(|_| KeyUpdatePrivate::random(rng))
+            .collect::<Vec<_>>();
+        assert_ne!(KeyUpdatePrivate::sum(&shares).x, BigNumber::zero());
```

Comment:
There is a negligible chance that this error case will be triggered (depending on the randomness used in the test algorithm). Is that okay?

Reply:
Then that would more likely be a bug than the 1/2^256 chance. It's supposed to be a real RNG in the test.

This is to exercise the sum function and show the contrast with the "equal zero" below.